### ISSUE TITLE
PM UI:  not automatically closed when project unloaded/removed

### ISSUE THREAD

**Issue by dtivel** [NuGet/Home#9935]
If PM UI is open for a project when you remove the project from Solution Manager, PM UI will no longer be closeable.

1.  In a remote session, open a solution with a project.
2.  Open PM UI for that project.
3.  In Solution Explorer right-click on the project and select Remove.

Expected result:  when the project is removed, PM UI should automatically close.

Actual result:  PM UI does not automatically close.  Also, PM UI cannot be closed at this point.

**Comment by rrelyea** [comment]
analysis so far:
solution close causes our toolwindow onclose method to be called, but we hit an exception in remoting code while saving settings.
also, we need to handle project/removed and project/unloaded cases -- events we've never handled before.

**Comment by rrelyea** [comment]
merged into dev-rrelyea-closepmui the work to date: https://github.com/NuGet/NuGet.Client/commit/57ef4ba3c49943e4cf1758ea431ccff939db3890
1) why is the exception happening? workaround for now?
2) i've started wiring to listen to beforeprojectunloaded event which is available from ivssolutionevents3, if i recall. this part of the work isn't compiling.
3) once 2 is done, listen to project removed and project unloaded, and close the windowframe ourselves, is the approach i was heading towards.

**Comment by dtivel** [comment]
There are multiple bugs here:
1.  One bug fixed on the VS side.  With the fix clients should receive the document close event when the associated project closes.
2.  It's unclear how to associate the PM UI document window with a project.
3.  When PM UI is closed manually, an exception prevents the window from actually closing.

**Comment by dtivel** [comment]
https://github.com/NuGet/NuGet.Client/commit/c219b612c2e09ca8bf0f801c833b3ebbe4dfabb5 unblocked progress on this.

**Comment by dtivel** [comment]
This is currently blocked on being able to obtain an `IVsUIHierarchy` object for the selected project in a Codespaces-connected client.  See AzDO bug 1264138.

**Comment by dtivel** [comment]
The fix for AzDO 1264138 has integrated with VS.  This is now unblocked.

**Comment by aortiz-msft** [comment]
What's missing here is that the PM UI won't close automatically if unloading or removing the project.

**Comment by donnie-msft** [comment]
Did not repro in VS 17.10 (without CodeSpaces).

Closing CodeSpaces issues.
